cmd/kubeadm/app/phases/uploadconfig/uploadconfig_test.go

```diff
@@ -19,37 +19,79 @@ package uploadconfig
+	core "k8s.io/client-go/testing"
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
 )
 
 func TestUploadConfiguration(t *testing.T) {
 	tests := []struct {
-		name    string
-		cfg     *kubeadmapi.MasterConfiguration
-		wantErr bool
+		name           string
+		createErr      error
+		updateErr      error
+		idempotentTest bool
+		wantErr        bool
+		verifyResult   bool
 	}{
 		{
-			"basic validation with correct key",
-			&kubeadmapi.MasterConfiguration{
-				KubernetesVersion: "1.7.3",
-			},
-			false,
+			name:         "basic validation with correct key",
+			wantErr:      false,
+			verifyResult: true,
+		},
+		{
+			name:           "update existing should report no error",
+			idempotentTest: true,
+			wantErr:        false,
+			verifyResult:   true,
+		},
+		{
+			name:      "unexpected errors for create should be returned",
+			createErr: apierrors.NewUnauthorized(""),
+			wantErr:   true,
+		},
+		{
+			name:           "unexpected errors for update should be returned",
+			updateErr:      apierrors.NewUnauthorized(""),
+			idempotentTest: true,
+			wantErr:        true,
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			cfg := &kubeadmapi.MasterConfiguration{
+				KubernetesVersion: "1.7.3",
+			}
 			client := clientsetfake.NewSimpleClientset()
-			if err := UploadConfiguration(tt.cfg, client); (err != nil) != tt.wantErr {
+			if tt.createErr != nil {
+				client.PrependReactor("create", "configmaps", func(action core.Action) (bool, runtime.Object, error) {
+					return true, nil, tt.createErr
+				})
+			}
+			// For idempotent test, we check the result of the second call.
+			if err := UploadConfiguration(cfg, client); !tt.idempotentTest && ((err != nil) != tt.wantErr) {
 				t.Errorf("UploadConfiguration() error = %v, wantErr %v", err, tt.wantErr)
 			}
-			masterCfg, err := client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(kubeadmconstants.MasterConfigurationConfigMap, metav1.GetOptions{})
-			if err != nil {
-				t.Errorf("Fail to query ConfigMap error = %v", err)
-			} else if masterCfg.Data[kubeadmconstants.MasterConfigurationConfigMapKey] == "" {
-				t.Errorf("Fail to find ConfigMap key = %v", err)
+			if tt.idempotentTest {
+				if tt.updateErr != nil {
+					client.PrependReactor("update", "configmaps", func(action core.Action) (bool, runtime.Object, error) {
+						return true, nil, tt.updateErr
+					})
+				}
+				if err := UploadConfiguration(cfg, client); (err != nil) != tt.wantErr {
+					t.Errorf("UploadConfiguration() error = %v", err)
+				}
+			}
+			if tt.verifyResult {
+				masterCfg, err := client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(kubeadmconstants.MasterConfigurationConfigMap, metav1.GetOptions{})
+				if err != nil {
+					t.Errorf("Fail to query ConfigMap error = %v", err)
+				} else if masterCfg.Data[kubeadmconstants.MasterConfigurationConfigMapKey] == "" {
```

Comment:
This verifies that the configmap can be found, but doesn't verify that its contents are in the expected format or are correct. How about we deserialize them and make sure they match the `MasterConfiguration` we passed in?

https://github.com/kubernetes/kubernetes/blob/d725fe2c2c17d4e2cd281384ebfaf4ff3800a8c3/cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go#L54